osh/cmd_eval.py

```diff
@@ -225,6 +225,54 @@ def PlusEquals(old_val, val):
     return val
 
 
+class Func(vm._Callable):
+
+    def __init__(self, node, mem, cmd_ev):
+        # type: (command.Func, state.Mem, CommandEvaluator) -> None
+        self.name = lexer.TokenVal(node.name)
+        self.node = node
+        self.cmd_ev = cmd_ev
+        self.mem = mem
+
+    def Call(self, pos_args, named_args):
+        # type: (List[value_t], Dict[str, value_t]) -> value_t
+        nargs = len(pos_args)
+        expected = len(self.node.pos_params)
+        if nargs != expected:
+            raise error.InvalidType("%s() expects %d arguments but %d were given" % (self.name, expected, nargs), self.node.keyword)
+
+        nargs = len(named_args)
+        expected = len(self.node.named_params)
+        if nargs != expected:
+            raise error.InvalidType("%s() expects %d named arguments but %d were given" % (self.name, expected, nargs), self.node.keyword)
+
+        with state.ctx_FuncCall(self.cmd_ev.mem, self):
+            for i in xrange(0, len(pos_args)):
+                pos_arg = pos_args[i]
+                pos_param = self.node.pos_params[i]
+
+                arg_name = location.LName(lexer.TokenVal(pos_param.name))
+
+                self.mem.SetLocationToken(pos_param.name)
```

Comment:
I think we don't need this, because SetLocationToken() is only for generating error messages when evaluating, and we're not evaluating things in between

(although I suppose when we add default arg passing we might)